lib/runtime/procs/read.js

```diff
@@ -0,0 +1,282 @@
+            if (result.points && result.points.length !== 0) {
+                this.logger.debug('read returned', result.points.length, 'points');
+                this.enqueue(result.points);
+            } else {
+                this.logger.debug('read returned no points');
+            }
+
+            this.readState = result.state;
+
+            // Check if we are being torn down.
+            if (this.inTeardown) {
+                return;
+            }
+
+            // The adapter indicates that it's completed reading all points up
+            // to the given end time by returning the ending time in readEnd.
+            //
+            // Notify the base class so it can emit any necessary ticks and
+            // check if we're all done, or call us back when it is time to
+            // reschedule.
+            //
+            // Otherwise we need to reschedule another call.
+            if (result.readEnd) {
+                this.advance(result.readEnd);
+            } else {
+                this.moreToRead = true;
+
+                // XXX this should be done by the scheduler once it drains the
+                // buffer.
+                this.schedule();
+            }
+        })
+        .catch((err) => {
+            this.logger.debug('read returned error', err.toString());
+            this.trigger('error', err);
+            this.emit_eof();
+        });
+    },
+
+    // Based on whether or not the current read is completed, update the time
+    // bounds and schedule another read.
+    //
+    // XXX this will need to be reworked to interact with the flow control
+    // scheduler.
+    schedule: function() {
+        if (! this.moreToRead) {
+            // If to is not :end: then this is a purely historical read, so once
+            // it is done with the initial batch, we can simply emit eof and
+            // be done.
+            if (! this.to.isEnd()) {
+                this.logger.debug('read done for historical source. emitting eof.');
+                this.emit_eof();
+                return;
+            }
+
+            // Otherwise, advance the readFrom and readTo bounds and schedule a
+            // callback for the next time we should read.
+            this.readFrom = this.readTo;
+            this.readTo = this.readFrom.add(this.readEvery);
+            var nextRead = this.readTo.add(this.lag);
+            this.logger.debug('scheduling next read at', nextRead.toJSON());
+            this.program.scheduler.schedule(nextRead.unixms(), _.bind(this.run, this));
+        } else {
+            // There's more to read in the current time period, so ask the
+            // scheduler to call us back and read more.
+            this.program.scheduler.schedule(this.program.scheduler.now(), _.bind(this.run, this));
+        }
+    },
+
+    teardown: function() {
+        if (! this.reading) {
+            this.emit_eof();
+        } else {
+            this.reading.then(() => {
+                this.emit_eof();
```

Comment:
Good point -- that doesn't seem like a good thing.
Even worse, I noticed that if there is a pending call into the scheduler that never gets cancelled or cleaned up.